File: app/src/main/java/com/example/customviewtestproject/customView/DashedCircleProgressPainter.java
package com.example.customviewtestproject.customView;

import android.graphics.*;

public class DashedCircleProgressPainter implements Painter {

    private RectF progressCircle;
    private Paint progressPaint;
    private Paint progressDotPaint;
    private Paint textPaint;
    private int textSize;
    private int color;
    private int plusAngle = 0;
    private int strokeWidth;
    private float dotRadius;
    private int min;
    private int max;
    private int width;
    private int height;
    private int padding;
    private int segmentsCount;

    DashedCircleProgressPainter(int color, int textSize, int min, int max, int progressStrokeWidth, int defaultPadding, int segmentsCount) {
        this.color = color;
        this.textSize = textSize;
        this.min = min;
        this.max = max;
        this.strokeWidth = progressStrokeWidth;
        this.padding = defaultPadding;
        this.segmentsCount = segmentsCount;
        init();
    }

    private void init() {
        progressPaint = new Paint();
        progressPaint.setAntiAlias(true);
        progressPaint.setStrokeWidth(strokeWidth);
        progressPaint.setColor(color);
        progressPaint.setStyle(Paint.Style.STROKE);
        int dashWith = 5;
        int dashSpace = 8;
        progressPaint.setPathEffect(new DashPathEffect(new float[]{dashWith, dashSpace},
                dashSpace));

        progressDotPaint = new Paint();
        progressDotPaint.setAntiAlias(true);
        progressDotPaint.setStrokeCap(Paint.Cap.ROUND);
        int dotStrokeWidth = strokeWidth;
        progressDotPaint.setStrokeWidth(dotStrokeWidth);
        progressDotPaint.setColor(color);
        progressDotPaint.setStyle(Paint.Style.FILL_AND_STROKE);

        textPaint = new Paint();
        textPaint.setColor(Color.WHITE);
        textPaint.setStyle(Paint.Style.FILL);
        textPaint.setTextAlign(Paint.Align.CENTER);
        textPaint.setTextSize(textSize);
    }

    private void initDashedCircleProgress() {
        progressCircle = new RectF();
        float halfOffset = strokeWidth / 2f;
        progressCircle.left = halfOffset + strokeWidth;
        progressCircle.top = halfOffset + strokeWidth;
        progressCircle.right = width - halfOffset - strokeWidth;
        progressCircle.bottom = height - halfOffset - strokeWidth;
        dotRadius = progressCircle.width() / 2f;
    }

    @Override
    public void draw(Canvas canvas) {
        int startAngle = 0;
        canvas.drawArc(progressCircle, startAngle, plusAngle, false, progressPaint);
        double angleRadians = Math.toRadians(startAngle + plusAngle + 180);
        float cos = (float) Math.cos(angleRadians);
        float sin = (float) Math.sin(angleRadians);
        float x = progressCircle.centerX() - dotRadius * cos;
        float y = progressCircle.centerY() - dotRadius * sin;
        canvas.drawPoint(x, y, progressDotPaint);

        String text = plusAngle + "°";
        canvas.drawText(text, progressCircle.centerX() - (textPaint.descent() + textPaint.ascent()) / 2 * textPaint.getTextSkewX(), progressCircle.centerY() - ((textPaint.descent() + textPaint.ascent()) / 2), textPaint);
    }

    public float getMin() {
        return min;
    }

    public void setMin(int min) {
        this.min = min;
    }

    public float getMax() {
        return max;
    }

    public void setMax(int max) {
        this.max = max;
    }

    public void setValue(int value) {
        this.plusAngle = (value * segmentsCount) % 360;
    }

    @Override
    public void onSizeChanged(int height, int width) {
        this.width = width - padding;
        this.height = height - padding;
        initDashedCircleProgress();
    }

    public int getColor() {
        return color;
    }

    public void setColor(int color) {
        this.color = color;
        progressPaint.setColor(color);
        progressDotPaint.setColor(color);
    }
}

File: app/src/main/java/com/example/customviewtestproject/customView/DashedCircleProgressPainter.kt
package com.example.customviewtestproject.customView

import android.graphics.*

class DashedCircleProgressPainter(
    private var color: Int,
    private val textSize: Int,
    private var min: Int,
    private var max: Int,
    private val strokeWidth: Int,
    private val padding: Int,
    private val segmentsCount: Int
) : Painter {

    private var progressCircle: RectF? = null
    private var progressPaint: Paint? = null
    private var progressDotPaint: Paint? = null
    private var horizontalLinePaint: Paint? = null
    private var progressHorizontalPaint: Paint? = null
    private var textAnglePaint: Paint? = null
    private var textLevelPaint: Paint? = null
    private var plusAngle = 0
    private var level = 0
    private var dotRadius: Float = 0.toFloat()
    private var width: Int = 0
    private var height: Int = 0

    init {
        progressPaint = Paint()
        progressPaint!!.isAntiAlias = true
        progressPaint!!.strokeWidth = strokeWidth.toFloat()
        progressPaint!!.color = color
        progressPaint!!.style = Paint.Style.STROKE
        val dashWith = 5
        val dashSpace = 8
        progressPaint!!.pathEffect = DashPathEffect(
            floatArrayOf(dashWith.toFloat(), dashSpace.toFloat()),
            dashSpace.toFloat()
        )

        progressDotPaint = Paint()
        progressDotPaint!!.isAntiAlias = true
        progressDotPaint!!.strokeCap = Paint.Cap.ROUND
        val dotStrokeWidth = strokeWidth - dashWith
        progressDotPaint!!.strokeWidth = dotStrokeWidth.toFloat()
        progressDotPaint!!.color = color
        progressDotPaint!!.style = Paint.Style.FILL_AND_STROKE

        textAnglePaint = Paint()
        textAnglePaint!!.color = Color.WHITE
        textAnglePaint!!.style = Paint.Style.FILL
        textAnglePaint!!.textAlign = Paint.Align.CENTER
        textAnglePaint!!.textSize = textSize.toFloat()

        horizontalLinePaint = Paint()
        horizontalLinePaint!!.isAntiAlias = true
        horizontalLinePaint!!.strokeCap = Paint.Cap.ROUND
        horizontalLinePaint!!.strokeWidth = dotStrokeWidth / 2f
        horizontalLinePaint!!.color = Color.DKGRAY
        horizontalLinePaint!!.style = Paint.Style.FILL_AND_STROKE

        progressHorizontalPaint = Paint()
        progressHorizontalPaint!!.isAntiAlias = true
        progressHorizontalPaint!!.strokeCap = Paint.Cap.ROUND
        progressHorizontalPaint!!.strokeWidth = dotStrokeWidth / 2f
        progressHorizontalPaint!!.color = color
        progressHorizontalPaint!!.style = Paint.Style.FILL_AND_STROKE

        textLevelPaint = Paint()
        textLevelPaint!!.color = color
        textLevelPaint!!.style = Paint.Style.FILL
        textLevelPaint!!.textAlign = Paint.Align.CENTER
        textLevelPaint!!.textSize = textSize.toFloat() / 4
    }

    private fun initDashedCircleProgress() {
        progressCircle = RectF()
        val halfOffset = strokeWidth / 2f
        progressCircle!!.left = halfOffset + strokeWidth
        progressCircle!!.top = halfOffset + strokeWidth
        progressCircle!!.right = width.toFloat() - halfOffset - strokeWidth.toFloat()
        progressCircle!!.bottom = height.toFloat() - halfOffset - strokeWidth.toFloat()
        dotRadius = progressCircle!!.width() / 2f
    }

    override fun draw(canvas: Canvas) {
        val startAngle = 0
        canvas.drawArc(progressCircle!!, startAngle.toFloat(), plusAngle.toFloat(), false, progressPaint!!)
        val angleRadians = Math.toRadians((startAngle + plusAngle + 180).toDouble())
        val cos = Math.cos(angleRadians).toFloat()
        val sin = Math.sin(angleRadians).toFloat()
        val x = progressCircle!!.centerX() - dotRadius * cos
        val y = progressCircle!!.centerY() - dotRadius * sin
        canvas.drawPoint(x, y, progressDotPaint!!)

        val textAngle = "$plusAngle°"
        canvas.drawText(
            textAngle,
            progressCircle!!.centerX() - (textAnglePaint!!.descent() + textAnglePaint!!.ascent()) / 2 * textAnglePaint!!.textSkewX,
            progressCircle!!.centerY() - (textAnglePaint!!.descent() + textAnglePaint!!.ascent()) / 2,
            textAnglePaint!!
        )
        canvas.drawLine(
            progressCircle!!.centerX() - progressCircle!!.centerX() / 3f,
            progressCircle!!.centerY() - progressCircle!!.centerY() / 3f,
            progressCircle!!.centerX() + progressCircle!!.centerX() / 3f,
            progressCircle!!.centerY() - progressCircle!!.centerY() / 3f,
            horizontalLinePaint!!
        )

        val startX = progressCircle!!.centerX() - progressCircle!!.centerX() / 3f
        val stopX = progressCircle!!.centerX() + progressCircle!!.centerX() / 3f

        canvas.drawLine(
            startX,
            progressCircle!!.centerY() - progressCircle!!.centerY() / 3f,
            startX + stopX * plusAngle / max.toFloat() / (stopX / startX),
            progressCircle!!.centerY() - progressCircle!!.centerY() / 3f,
            progressHorizontalPaint!!
        )

        val textLevel = "LEVEL $level"
        canvas.drawText(
            textLevel,
            progressCircle!!.centerX() - (textLevelPaint!!.descent() + textLevelPaint!!.ascent()) / 2 * textLevelPaint!!.textSkewX,
            progressCircle!!.centerY() - progressCircle!!.centerY() / 3f - 2f * strokeWidth - (textLevelPaint!!.descent() + textLevelPaint!!.ascent()),
            textLevelPaint!!
        )
    }

    fun getMin(): Float {
        return min.toFloat()
    }

    fun setMin(min: Int) {
        this.min = min
    }

    fun getMax(): Float {
        return max.toFloat()
    }

    fun setMax(max: Int) {
        this.max = max
    }

    fun setValue(value: Int) {
        level = calculateLevel(value)
        plusAngle = calculateAngle(value)
    }

    private fun calculateLevel(value: Int): Int {
        var level = 0
        if (value > 0) {
            level = value / (max / segmentsCount)
        }
        if (level >= segmentsCount) {
            return level
        }
        level++
        return level
    }

    private fun calculateAngle(value: Int): Int {
        var calculatedAngle = value * segmentsCount % max
        if (value > 0 && calculatedAngle == 0) {
            calculatedAngle = max
        }
        return calculatedAngle
    }

    override fun onSizeChanged(height: Int, width: Int) {
        this.width = width - padding
        this.height = height - padding
        initDashedCircleProgress()
    }

    override fun getColor(): Int {
        return color
    }

    override fun setColor(color: Int) {
        this.color = color
        progressPaint!!.color = color
        progressDotPaint!!.color = color
    }
}
